# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsRange.h
# Commit: b1f1e6ad93f8
# Full Hash: b1f1e6ad93f8d7634acfcd0d66467c6efc2ad80e
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2024-03-26 04:11:24
# Regressor Bug: 1867058
# File Overlap Count: 1
# Description:
#   Bug 1867058 - Part 5: Update nsRange to create and update mCrossBoundaryRange when necessary r=smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D195305
# ==============================================================================

diff -r 02d89b880ac5 -r b1f1e6ad93f8 dom/base/nsRange.h
--- a/dom/base/nsRange.h	Mon Mar 25 13:40:57 2024 +0000
+++ b/dom/base/nsRange.h	Mon Mar 25 13:40:57 2024 +0000
@@ -32,6 +32,13 @@
 class DOMRectList;
 class InspectorFontFace;
 class Selection;
+
+enum class CollapsePolicy : uint8_t {
+  No,                                       // Don't need to collapse
+  DefaultRange,                             // Collapse the default range
+  DefaultRangeAndCrossShadowBoundaryRanges  // Collapse both the default range
+                                            // and the cross boundary range
+};
 }  // namespace dom
 }  // namespace mozilla
 
@@ -44,6 +51,8 @@
   using DOMRectList = mozilla::dom::DOMRectList;
   using RangeBoundary = mozilla::RangeBoundary;
   using RawRangeBoundary = mozilla::RawRangeBoundary;
+  using AllowRangeCrossShadowBoundary =
+      mozilla::dom::AllowRangeCrossShadowBoundary;
 
   virtual ~nsRange();
   explicit nsRange(nsINode* aNode);
@@ -112,14 +121,20 @@
    * When you set both start and end of a range, you should use
    * SetStartAndEnd() instead.
    */
-  nsresult SetStart(nsINode* aContainer, uint32_t aOffset) {
+  nsresult SetStart(nsINode* aContainer, uint32_t aOffset,
+                    AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                        AllowRangeCrossShadowBoundary::No) {
     ErrorResult error;
-    SetStart(RawRangeBoundary(aContainer, aOffset), error);
+    SetStart(RawRangeBoundary(aContainer, aOffset), error,
+             aAllowCrossShadowBoundary);
     return error.StealNSResult();
   }
-  nsresult SetEnd(nsINode* aContainer, uint32_t aOffset) {
+  nsresult SetEnd(nsINode* aContainer, uint32_t aOffset,
+                  AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                      AllowRangeCrossShadowBoundary::No) {
     ErrorResult error;
-    SetEnd(RawRangeBoundary(aContainer, aOffset), error);
+    SetEnd(RawRangeBoundary(aContainer, aOffset), error,
+           aAllowCrossShadowBoundary);
     return error.StealNSResult();
   }
 
@@ -225,6 +240,11 @@
   void SetStartAfterJS(nsINode& aNode, ErrorResult& aErr);
   void SetStartBeforeJS(nsINode& aNode, ErrorResult& aErr);
 
+  void SetStartAllowCrossShadowBoundary(nsINode& aNode, uint32_t aOffset,
+                                        ErrorResult& aErr);
+  void SetEndAllowCrossShadowBoundary(nsINode& aNode, uint32_t aOffset,
+                                      ErrorResult& aErr);
+
   void SurroundContents(nsINode& aNode, ErrorResult& aErr);
   already_AddRefed<DOMRect> GetBoundingClientRect(bool aClampToEdge = true,
                                                   bool aFlushLayout = true);
@@ -236,14 +256,26 @@
   // Following methods should be used for internal use instead of *JS().
   void SelectNode(nsINode& aNode, ErrorResult& aErr);
   void SelectNodeContents(nsINode& aNode, ErrorResult& aErr);
-  void SetEnd(nsINode& aNode, uint32_t aOffset, ErrorResult& aErr);
-  void SetEnd(const RawRangeBoundary& aPoint, ErrorResult& aErr);
+  void SetEnd(nsINode& aNode, uint32_t aOffset, ErrorResult& aErr,
+              AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                  AllowRangeCrossShadowBoundary::No);
+  void SetEnd(const RawRangeBoundary& aPoint, ErrorResult& aErr,
+              AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                  AllowRangeCrossShadowBoundary::No);
   void SetEndAfter(nsINode& aNode, ErrorResult& aErr);
-  void SetEndBefore(nsINode& aNode, ErrorResult& aErr);
-  void SetStart(nsINode& aNode, uint32_t aOffset, ErrorResult& aErr);
-  void SetStart(const RawRangeBoundary& aPoint, ErrorResult& aErr);
+  void SetEndBefore(nsINode& aNode, ErrorResult& aErr,
+                    AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                        AllowRangeCrossShadowBoundary::No);
+  void SetStart(nsINode& aNode, uint32_t aOffset, ErrorResult& aErr,
+                AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                    AllowRangeCrossShadowBoundary::No);
+  void SetStart(const RawRangeBoundary& aPoint, ErrorResult& aErr,
+                AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                    AllowRangeCrossShadowBoundary::No);
   void SetStartAfter(nsINode& aNode, ErrorResult& aErr);
-  void SetStartBefore(nsINode& aNode, ErrorResult& aErr);
+  void SetStartBefore(nsINode& aNode, ErrorResult& aErr,
+                      AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary =
+                          AllowRangeCrossShadowBoundary::No);
   void Collapse(bool aToStart);
 
   static void GetInnerTextNoFlush(mozilla::dom::DOMString& aValue,
@@ -352,6 +384,13 @@
    */
   nsINode* GetRegisteredClosestCommonInclusiveAncestor();
 
+  template <typename SPT, typename SRT, typename EPT, typename ERT>
+  void CreateOrUpdateCrossShadowBoundaryRangeIfNeeded(
+      const mozilla::RangeBoundaryBase<SPT, SRT>& aStartBoundary,
+      const mozilla::RangeBoundaryBase<EPT, ERT>& aEndBoundary);
+
+  void ResetCrossShadowBoundaryRange() { mCrossShadowBoundaryRange = nullptr; }
+
   /*
    * The methods marked with MayCrossShadowBoundary[..] additionally check for
    * the existence of mCrossShadowBoundaryRange, which indicates a range that
@@ -361,13 +400,13 @@
    * which provide a slightly faster implementation.
    * */
 
-  nsIContent* GetMayCrossBoundaryChildAtStartOffset() const {
+  nsIContent* GetMayCrossShadowBoundaryChildAtStartOffset() const {
     return mCrossShadowBoundaryRange
                ? mCrossShadowBoundaryRange->GetChildAtStartOffset()
                : mStart.GetChildAtOffset();
   }
 
-  nsIContent* GetMayCrossBoundaryChildAtEndOffset() const {
+  nsIContent* GetMayCrossShadowBoundaryChildAtEndOffset() const {
     return mCrossShadowBoundaryRange
                ? mCrossShadowBoundaryRange->GetChildAtEndOffset()
                : mEnd.GetChildAtOffset();
@@ -377,34 +416,34 @@
     return mCrossShadowBoundaryRange;
   }
 
-  nsINode* GetMayCrossBoundaryStartContainer() const {
+  nsINode* GetMayCrossShadowBoundaryStartContainer() const {
     return mCrossShadowBoundaryRange
                ? mCrossShadowBoundaryRange->GetStartContainer()
                : mStart.Container();
   }
 
-  nsINode* GetMayCrossBoundaryEndContainer() const {
+  nsINode* GetMayCrossShadowBoundaryEndContainer() const {
     return mCrossShadowBoundaryRange
                ? mCrossShadowBoundaryRange->GetEndContainer()
                : mEnd.Container();
   }
 
-  uint32_t MayCrossBoundaryStartOffset() const {
+  uint32_t MayCrossShadowBoundaryStartOffset() const {
     return mCrossShadowBoundaryRange ? mCrossShadowBoundaryRange->StartOffset()
                                      : StartOffset();
   }
 
-  uint32_t MayCrossBoundaryEndOffset() const {
+  uint32_t MayCrossShadowBoundaryEndOffset() const {
     return mCrossShadowBoundaryRange ? mCrossShadowBoundaryRange->EndOffset()
                                      : EndOffset();
   }
 
-  const RangeBoundary& MayCrossBoundaryStartRef() const {
+  const RangeBoundary& MayCrossShadowBoundaryStartRef() const {
     return mCrossShadowBoundaryRange ? mCrossShadowBoundaryRange->StartRef()
                                      : StartRef();
   }
 
-  const RangeBoundary& MayCrossBoundaryEndRef() const {
+  const RangeBoundary& MayCrossShadowBoundaryEndRef() const {
     return mCrossShadowBoundaryRange ? mCrossShadowBoundaryRange->EndRef()
                                      : EndRef();
   }
@@ -430,7 +469,9 @@
   MOZ_CAN_RUN_SCRIPT_BOUNDARY void DoSetRange(
       const mozilla::RangeBoundaryBase<SPT, SRT>& aStartBoundary,
       const mozilla::RangeBoundaryBase<EPT, ERT>& aEndBoundary,
-      nsINode* aRootNode, bool aNotInsertedYet = false);
+      nsINode* aRootNode, bool aNotInsertedYet = false,
+      mozilla::dom::CollapsePolicy aCollapsePolicy = mozilla::dom::
+          CollapsePolicy::DefaultRangeAndCrossShadowBoundaryRanges);
 
   // Assume that this is guaranteed that this is held by the caller when
   // this is used.  (Note that we cannot use AutoRestore for mCalledByJS